### ISSUE TITLE
matplotlib is missing GSettings

### ISSUE THREAD

**Issue by rnhmjoj** [NixOS/nixpkgs#25351]
## Issue description
When opening the file picker of a matplotlib (gtk3 backend) window the process crashes with this error:
`GLib-GIO-ERROR **: No GSettings schemas are installed on the system`
I think it needs wrapGAppsHook but I'm not really sure how that works.

### Steps to reproduce
```
$ nix-shell -p 'pythonPackages.matplotlib.override {enableGtk3=true;}' --run 'python -c "import matplotlib.pyplot as plt; plt.plot([1,2,3],[4,5,6]); plt.show()"'
```

## Technical details

NixOS version: 17.09pre106117.c90998d5cf

**Comment by rnhmjoj** [comment]
Reading https://github.com/NixOS/nixpkgs/issues/24880 I tried adding to the buildInputs `wrapGAppsHook` and `gsettings_desktop_schemas` but it didn't help. Anyone knows how to fix it?

**Comment by mmahut** [comment]
Any news on this issue?

**Comment by rnhmjoj** [comment]
Nothing new: It's still happening. Also the package probably has to be wrapped for #65399 .
I have no idea how to to do that on a python library, though.

**Comment by doronbehar** [comment]
I've just solved this error in an end user application - https://github.com/NixOS/nixpkgs/pull/84449 The solution is to use `wrapGAppsHook` which adds something like this to the wrapper:

```
export XDG_DATA_DIRS='/nix/store/nwksqszw0dakmizrrlql9pamcf39vnxg-gtk+3-3.24.14/share/gsettings-schemas/gtk+3-3.24.14:/nix/store/slmsmbyfgf1rqj9rp6s1y33cqaz8q203-gsettings-desktop-schemas-3.34.0/share/gsettings-schemas/gsettings-desktop-schemas-3.34.0:/nix/store/nwksqszw0dakmizrrlql9pamcf39vnxg-gtk+3-3.24.14/share/gsettings-schemas/gtk+3-3.24.14:/nix/store/slmsmbyfgf1rqj9rp6s1y33cqaz8q203-gsettings-desktop-schemas-3.34.0/share/gsettings-schemas/gsettings-desktop-schemas-3.34.0'${XDG_DATA_DIRS:+':'}$XDG_DATA_DIRS
export XDG_DATA_DIRS='/nix/store/92p57svcsqgwa5fgb8z9qg7m7r66nx7a-guvcview-2.0.6/share'${XDG_DATA_DIRS:+':'}$XDG_DATA_DIRS
```

To solve this issue in the context of using matplotlib is somewhat challenging because matplotlib is a library - what exactly would we wrap? https://github.com/NixOS/nixpkgs/pull/83321 is exactly what we need.

**Comment by rnhmjoj** [comment]
@doronbehar Sadly there is no way, yet. The situation will hopefully change once structured attributes are enabled in nixpkgs and declarative wrappers are implemented. See this issue #72074.

> what exactly would we wrap? #83321 is exactly what we need.

Yes, I tried implementing a temporary solution with #83321 because this will take a while before it's usable. It needs some feedback and a lot of testing but it should be working.

**Comment by doronbehar** [comment]
Still an issue.

**Comment by rnhmjoj** [comment]
Still an issue.

**Comment by rnhmjoj** [comment]
Still an issue, sadly.

**Comment by rnhmjoj** [comment]
Not really, it's possible to fix these issues without a unified wrapping mechanism. See https://github.com/NixOS/nixpkgs/pull/102949#issuecomment-1587237994

**Comment by Artturin** [comment]
some patching to make XDG_DATA_DIRS contain schemas before https://github.com/matplotlib/matplotlib/blob/a861b8af82cc4618e0a631e5eac5b4030b3f04a1/lib/matplotlib/backends/backend_gtk3.py#L340 could help

**Comment by jtojnar** [comment]
I think the simplest solution would be hardcoding the schema path in GTK itself. We now have semi-automated patch generating tooling (see evolution-data-server package for an example), I just have not gotten around to applying it to GTK yet.
